### PR TITLE
NMS-12392: Fix send-event.pl with time zone support

### PR DESCRIPTION
**Pull request by indigo423**
### All Contributors

* [x] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [x] Have you made a comment in that issue which points back to this PR?
* [x] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [x] If this is a new or updated feature, is there documentation for the new behavior?
* [x] If this is new code, are there unit and/or integration tests?
* [x] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

This PR applies a patch from attached to the issue to fix our send-event.pl script which got busted when we introduced time zone changes.

With all the kudos to the contributor: @jb4415🏅

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12392

### CODE REVIEW

File: opennms-base-assembly/src/main/filtered/bin/send-event.pl [hunk 1]
```diff
@@ -187,7 +188,7 @@ END
 $event .= "   <nodeid>$NODEID</nodeid>\n"          if (defined $NODEID);
 
 $event .= <<END;
-   <time>${year}-${month}-${mday}T${hour}:${min}:${sec}+00:00</time>
+   <time>$week[$wday], $mday $month[$month] $year $hour:$min:$sec o'clock $ZONE</time>
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/bin/send-event.pl]: We should continue using ISO-8601, and not revert to the old format.

**Comment by RangerRick** [comment]
Oh, hah, @j-white beat me to it.  :)

**Comment by indigo423** [comment]
> Oh, hah, @j-white beat me to it. :)

yeah my perl skills are extremely limited :D

**Comment by jb4415** [comment]
Timezone wasn't the issue that I reported. The value being inserted between the time tags was the correct GMT time for events.

The issue was the hour component of that time was not being zero padded so that if I submitted an event between midnight and 9:59 am GMT, the hour value only had a single digit.

My fix was to add a sprintf statement for the hour field to ensure the hour was always 2 digits. Similar to the other values that already have the sprintf command.

File: opennms-base-assembly/src/main/filtered/bin/send-event.pl [hunk 2]
```diff
@@ -163,11 +163,12 @@
...
 $year += 1900;
-my $month = sprintf("%02d", $mon + 1);
-$mday     = sprintf("%02d", $mday);
-$min      = sprintf("%02d", $min);
-$sec      = sprintf("%02d", $sec);
-
+my $month = $mon;
+$hour = sprintf("%02d", $hour);
+$min = sprintf("%02d", $min);
+$sec = sprintf("%02d", $sec);
```
Comment (jb4415) [inline opennms-base-assembly/src/main/filtered/bin/send-event.pl]: This is the only change that is required for this.  line 164 gets gmtime so there should be no need to be concerned about timezones as internally, the script is working in GMT.

**Comment by RangerRick** [comment]
I went ahead and reverted the  `c06dcdaa1ad48b488dbfe1ffe55607d1d5a61111` commit since the only thing necessary was the hours fix.  The zone stuff is wrong, because the code uses `gmtime()` so it's always UTC, thus the offset in the ISO bit is always 0.

If we _did_ want to implement actual time zone support, it would require different code than the bit from the old `send-event.pl` that was not really working (which is why I removed it when I made the original ISO changes).

**Comment by RangerRick** [comment]
What is the `incomplete` tag for?

**Comment by indigo423** [comment]
> What is the `incomplete` tag for?

Don't merge the PR, there is stuff missing and the PR is also not complete to be reviewed. You can remove the tag if this is not the case anymore.

**Comment by RangerRick** [comment]
I guess I'm just confused about what's missing.  The timezone stuff was a red herring, timezone support (that came back when you copied over some other `send-event.pl`) was _removed_ when I switched the script to support ISO-8601, because it was incomplete/incorrect in the first place.  The original issue is _only_ about fixing the format of the event sent.

**Comment by RangerRick** [comment]
(The diff in his original comment on the ticket was the diff between `25.0.0` and `25.1.0`, @jb4415 was only including it in the ticket to show the changes, AFAICT.)

**Comment by jb4415** [comment]
> (The diff in his original comment on the ticket was the diff between `25.0.0` and `25.1.0`, @jb4415 was only including it in the ticket to show the changes, AFAICT.)

Yes you're quite right. The diff I included was between the 25.0.0 and 25.1.0 versions. I was just trying to show that I could see that send-event.pl had been updated to reflect the new event XML format. I posted that before i figured out the fix. My fix was just to add the 0 padding of the hour field. Sorry for the confusion.

**Pull request merged**